File: app/src/androidTest/java/com/simplemobiletools/calculator/ScientificActivityTest.java
package com.simplemobiletools.calculator;

import androidx.test.rule.ActivityTestRule;
import androidx.test.runner.AndroidJUnit4;

import com.simplemobiletools.calculator.activities.MainActivity;
import com.simplemobiletools.calculator.activities.ScientificActivity;

import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.longClick;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;

@RunWith(AndroidJUnit4.class)
/**
 * Created by Ahmed Ashraf Hamza with Ahmad Osama on 2018-12-18.
 */

public class ScientificActivityTest {

    @Rule public final ActivityTestRule<ScientificActivity> activity = new ActivityTestRule<>(ScientificActivity.class);


    @Test
    public void SinTest() {
        press(R.id.btn_9);
        press(R.id.btn_0);
        press(R.id.btn_sin);
        checkResult("1");
        checkFormula("Sin(90)");
    }
    @Test
    public void CosTest() {
        press(R.id.btn_9);
        press(R.id.btn_0);
        press(R.id.btn_cos);
        checkResult("0");
        checkFormula("Cos(90)");
    }
    @Test
    public void TanTest() {
        press(R.id.btn_4);
        press(R.id.btn_5);
        press(R.id.btn_tan);
        checkResult("1");
        checkFormula("Tan(45)");
    }
    @Test
    public void ArcsinTest() {
        press(R.id.btn_1);
        press(R.id.btn_arcsin);
        checkResult("90");
        checkFormula("Arcsin(1)");
    }
    @Test
    public void ArccosTest() {
        press(R.id.btn_0);
        press(R.id.btn_arccos);
        checkResult("90");
        checkFormula("Arccos(0)");
    }
    @Test
    public void ArcTanTest() {
        press(R.id.btn_1);
        press(R.id.btn_arctan);
        checkResult("45");
        checkFormula("Arctan(1)");
    }


    private void press(int id) {
        onView(withId(id)).perform(click());
    }

    private void longPress(int id) {
        onView(withId(id)).perform(longClick());
    }

    private void checkResult(String desired) {
        onView(withId(R.id.result)).check(matches(withText(desired)));
    }

    private void checkFormula(String desired) {
        onView(withId(R.id.formula)).check(matches(withText(desired)));
    }

}

File: app/src/test/java/com/simplemobiletools/calculator/ScientificActivityTest.kt
package com.simplemobiletools.calculator

import com.simplemobiletools.calculator.activities.MainActivity
import com.simplemobiletools.calculator.activities.ScientificActivity
import com.simplemobiletools.calculator.helpers.*
import junit.framework.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.Robolectric
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
@Config(constants = BuildConfig::class, sdk = intArrayOf(21))
class ScientificActivityTest {
    private lateinit var activity: ScientificActivity

    private fun getDisplayedNumber() = activity.calc.displayedNumber

    @Before
    fun setUp() {
        activity = Robolectric.setupActivity(ScientificActivity::class.java)
    }



    @Test
    fun SinTest() {
        setDouble(90.0)
        handleOperation(Sin)
        assertEquals("1", getDisplayedNumber())
        checkFormula("Sin(90)")
    }
    @Test
    fun CosTest() {
        setDouble(90.0)
        handleOperation(Cos)
        assertEquals("0", getDisplayedNumber())
        checkFormula("Cos(90)")
    }
    @Test
    fun TanTest() {
        setDouble(45.0)
        handleOperation(Tan)
        assertEquals("1", getDisplayedNumber())
        checkFormula("Tan(45)")
    }
    @Test
    fun ArcsinTest() {
        setDouble(1.0)
        handleOperation(Arcsin)
        assertEquals("90", getDisplayedNumber())
        checkFormula("Arcsin(1)")
    }
    @Test
    fun ArcosTest() {
        setDouble(0.0)
        handleOperation(Arccos)
        assertEquals("90", getDisplayedNumber())
        checkFormula("Arccos(0)")
    }
    @Test
    fun ArctanTest() {
        setDouble(1.0)
        handleOperation(Arctan)
        assertEquals("45", getDisplayedNumber())
        checkFormula("Arctan(1)")
    }




    private fun setDouble(d: Double) {
        activity.setValueDouble(d)
    }

    private fun handleOperation(operation: String) {
        activity.calc.handleOperation(operation)
    }

    private fun checkFormula(desired: String) {
        assertEquals(desired, activity.calc.displayedFormula)
    }

    private fun calcResult(baseValue: Double, operation: String, secondValue: Double): String? {
        setDouble(baseValue)
        handleOperation(operation)
        setDouble(secondValue)
        activity.calc.handleResult()
        return getDisplayedNumber()
    }
}
